api/utils/keys/policy.go

```diff
@@ -34,49 +34,110 @@ const (
+	case PrivateKeyPolicyHardwareKeyPIN:
+		return keyPolicy.isHardwareKeyPINVerified()
+	case PrivateKeyPolicyHardwareKeyTouchAndPIN:
+		return keyPolicy.isHardwareKeyTouchVerified() && keyPolicy.isHardwareKeyPINVerified()
 	}
-	return NewPrivateKeyPolicyError(p)
+
+	return false
+}
+
+// Deprecated in favor of IsRequiredPolicyMet.
+// TODO(Joerger): delete once reference in /e is replaced.
+func (requiredPolicy PrivateKeyPolicy) VerifyPolicy(keyPolicy PrivateKeyPolicy) error {
+	if !IsRequiredPolicyMet(requiredPolicy, keyPolicy) {
+		return NewPrivateKeyPolicyError(requiredPolicy)
+	}
+	return nil
 }
 
-// IsHardwareKeyVerified return true if this private key policy requires a hardware key.
-func (p PrivateKeyPolicy) IsHardwareKeyVerified() bool {
+// IsHardwareKeyPolicy return true if this private key policy requires a hardware key.
+func (p PrivateKeyPolicy) IsHardwareKeyPolicy() bool {
 	switch p {
-	case PrivateKeyPolicyHardwareKey, PrivateKeyPolicyHardwareKeyTouch:
+	case PrivateKeyPolicyHardwareKey,
+		PrivateKeyPolicyHardwareKeyTouch,
+		PrivateKeyPolicyHardwareKeyPIN,
+		PrivateKeyPolicyHardwareKeyTouchAndPIN:
 		return true
 	}
 	return false
 }
 
-// MFAVerified checks that meet this private key policy counts towards MFA verification.
+// MFAVerified checks that private keys with this key policy count as MFA verified.
+// Both Hardware key touch and pin are count as MFA verification.
 func (p PrivateKeyPolicy) MFAVerified() bool {
-	return p == PrivateKeyPolicyHardwareKeyTouch
+	return p.isHardwareKeyTouchVerified() || p.isHardwareKeyPINVerified()
+}
+
+func (p PrivateKeyPolicy) isHardwareKeyTouchVerified() bool {
+	switch p {
+	case PrivateKeyPolicyHardwareKeyTouch, PrivateKeyPolicyHardwareKeyTouchAndPIN:
+		return true
+	}
+	return false
+}
+
+func (p PrivateKeyPolicy) isHardwareKeyPINVerified() bool {
+	switch p {
+	case PrivateKeyPolicyHardwareKeyPIN, PrivateKeyPolicyHardwareKeyTouchAndPIN:
+		return true
+	}
+	return false
 }
 
 func (p PrivateKeyPolicy) validate() error {
 	switch p {
-	case PrivateKeyPolicyNone, PrivateKeyPolicyHardwareKey, PrivateKeyPolicyHardwareKeyTouch:
+	case PrivateKeyPolicyNone,
+		PrivateKeyPolicyHardwareKey,
+		PrivateKeyPolicyHardwareKeyTouch,
+		PrivateKeyPolicyHardwareKeyPIN,
+		PrivateKeyPolicyHardwareKeyTouchAndPIN:
 		return nil
 	}
 	return trace.BadParameter("%q is not a valid key policy", p)
 }
 
+// GetPolicyFromSet returns least restrictive policy necessary to meet the given set of policies.
+func GetPolicyFromSet(policies []PrivateKeyPolicy) (PrivateKeyPolicy, error) {
+	setPolicy := PrivateKeyPolicyNone
+	for _, policy := range policies {
+		if !IsRequiredPolicyMet(policy, setPolicy) {
+			if IsRequiredPolicyMet(setPolicy, policy) {
+				// Upgrade set policy to stricter policy.
+				setPolicy = policy
+			} else if IsRequiredPolicyMet(policy, PrivateKeyPolicyHardwareKeyTouchAndPIN) &&
+				IsRequiredPolicyMet(setPolicy, PrivateKeyPolicyHardwareKeyTouchAndPIN) {
```

Comment:
Is this ever false for the set of policies we have today?

Reply:
No, but this additional check was suggested in a previous comment to future proof us for future key policies. For example, if we ever introduce `web_session`  and `hardware_key_strict` (which would prevent web sessions without PIV) as possible requirements.

Reply:
I'd add a small comment, but otherwise seems fine to me.